WalletWasabi/BitcoinCore/P2pNode.cs

```diff
@@ -162,28 +163,26 @@ public void Disconnect()
 
 				try
 				{
-					node.Disconnect();
+					// Disconnection not waited here.
+					node.DisconnectAsync();
```

Comment:
Use a continuation here. Something like:
```c#
node.DisconnectAsync(theCancellationToketThatCancelsAfterXXseconds)
        .ContinueWith(t => Logger.Log("no idea"); TaskContinuationOptions.NotOnRanToCompletion)
        .ContinueWith(t => Logger.Log("Timeout"); TaskContinuationOptions.OnlyOnCanceled);
```

Reply:
`node.DisconnectAsync()` cannot take cancellation. It just starts the disconnection 

https://github.com/MetacoSA/NBitcoin/blob/974f9e855c068e5e681361fbb207069bacf0d0b1/NBitcoin/Protocol/Node.cs#L1114

Reply:
Ohh I see. What I don't like is the `while` loop for waiting. I think you could consider using the `EventAwaiter` and await for the `Disconnected` event.

Reply:
I think Lucas's suggestion makes a lot of sense here. The concept of waiting for an event is implemented so why not to use that.

Reply:
EventAwaiter can be used if the Event is using EventHandler. In the case of disconnect it is just: 

`public delegate void NodeEventHandler(Node node);`

Reply:
I added a custom funtion here, according to EventAwaiter